### ISSUE TITLE
consider generating URL strings from a URLPattern

### ISSUE THREAD

**Issue by wanderview** [whatwg/urlpattern#73]
One of the most frequent pieces of feedback we have gotten is that developers would like to generate URL and component strings from a URLPattern.  For example, something like:

```javascript
const pattern = new URLPattern({ pathname: '/product/:id' });
const my_pathname = pattern.generate('pathname', { id: '12345' });
// my_pathname is "/product/12345"
```

There is a discussions thread about this in #41.  I'm filing this issue to consider as a future enhancement.

**Comment by Jamesernator** [comment]
The [reference library](https://github.com/pillarjs/path-to-regexp#compile-reverse-path-to-regexp) does have some behaviour we would probably want to smooth other, especially given people are likely to use this for user input.

In particular path segments should probably be always validated and encoded by default, i.e. no attacks using slashes or relative path segments (`.`/`..`).

Also path-to-regexp docs don't specify any behaviour for non-named pattern segments, e.g. what would we do with `new URLPattern("/*.jpg")`, one option would be to just throw an error, alternatively they could be treated as numeric-indexed, although this is tricky in itself e.g.:

```js
const pattern = new URLPattern(`/foo/*`);
// Almost certainly "/foo/test" is the right behaviour here, as there's only
// a single pattern perhaps the array wrapper could be optional too
pattern.generate("pathname", ["test"]);

const pattern2 = new URLPattern("/images/*.jpg");
// Less clear as to which of these would be correct, it would make sense that people
// would want to include the full name, although others might expect that only the * is replaced
pattern2.generate("pathname", ["1.jpg"]);
pattern2.generate("pathname", ["1"]);
```

It also make more sense as a separate API as with the reference library, especially if not all syntax is supported for the reason above. i.e. `new URLTemplate("/product/:id").generate("pathname", { id: "12345" })`

**Comment by Haroenv** [comment]
Just would like to chime in that this is very important, and that I can’t imagine a case where you read urls into objects without also turning objects into urls. Search pages for example have lots of generated links (think filters, pagination), but also read the url into local state.

**Comment by jimmywarting** [comment]
I were just merely considering if a result could be mutable...
for example: Say that you are on `shopify.com/categories/computers` and want to navigate to games section `shopify.com/categories/games`

```js
const pattern = new URLPattern({ pathname: '/categories/:category' });
const result = pattern.exec(location.href)
result.pathname.groups.category = 'games'
location.href = result.toString()
```
Thinking this could be useful for a adv search with 2 way data binding or something.

doe a `generate´ function could work as equally good, maybe mutable url pattern result isn't desirable?

**Comment by Haroenv** [comment]
the pattern being mutable can make sense, as new URL is a precedence, but personally I would need to make a copy of the pattern for every change then, as you have many links with different state, as it's not the current state you want to show in the URL, but a possible future state

**Comment by guy-borderless** [comment]
Would be helpful for my use case. I use a URL pattern to 

1. determine type of page
2. extract information from the url
3. reconstruct the URL from extracted information

I'm sure this will be a common enough use case for URLPattern

**Comment by samuelstroschein** [comment]
This API is highly benefitial for i18n/localization where localized URLs have to be mapped back to the original (deLocalized) URL. Here are docs for the use case that should be accounted for https://inlang.com/m/gerre34r/library-inlang-paraglideJs/strategy#url.

**Comment by sisidovski** [comment]
I feel it's better to expose this API to the [URLPattern](https://urlpattern.spec.whatwg.org/#urlpattern) interface, rather than making it mutable with adding `toString()` in [URLPatternResult](https://urlpattern.spec.whatwg.org/#dictdef-urlpatternresult), which is currently defined as dictionary and doesn't have any method.

**Comment by sisidovski** [comment]
We'd like to make a progress on this proposal to resolve https://github.com/w3c/IFT/issues/259.

I've summarized expected inputs and outputs here. I think this is still not comprehensive, appreciate your any feedback there.
https://docs.google.com/document/d/1ca6geyHD40MfHkalgEv4AcBo9-rHK6CrAOxwcWor9WA/edit?usp=sharing

**Comment by samuelstroschein** [comment]
@sisidovski Look into [this code](https://github.com/opral/monorepo/blob/3d5166a50756d6ef3b885673202a208ede240c20/inlang/packages/paraglide/paraglide-js/src/compiler/runtime/localize-url.js#L242-L365) a custom generate function that is used for localization. Might reveal one or two insights.

**Comment by yoshisatoyanagisawa** [comment]
Suggested by @annevk in https://docs.google.com/document/d/1ca6geyHD40MfHkalgEv4AcBo9-rHK6CrAOxwcWor9WA/edit?disco=AAABhYZWjLY, but we might need to consider the way to generate a complete URLPattern instead of specifying the component.  Note that the complete URL is somewhat touched in https://github.com/whatwg/urlpattern/issues/73#issuecomment-1028873691.

Will it make sense to pass a record to generate to deal with the complete URL?
JavaScript code example might be something like:
```
const pattern = URLPattern("https://example.com/:path");
const result = await pattern.generate({'path': 'image/a.jpg'});
const generatedURL = result.toString();  // expecting "https://example.com/image/a.jpg"
```

What do you think?

**Comment by sisidovski** [comment]
Yeah it makes sense consider generating a full URL.

fyi, the reference library only supports  pathname to generate a string tho. https://github.com/pillarjs/path-to-regexp#compile-reverse-path-to-regexp

I'm wondering how we should handle the case if there are two named groups which have the same name?

```js
const pattern = new URLPattern('https://{:pattern.}?example.com/:pattern');
const result = await pattern.generate({'pattern': 'image/a.jpg'});
// => ???
```

It looks the current design doesn't have a capability to substitute two strings which have same names in a different component.

**Comment by yoshisatoyanagisawa** [comment]
> fyi, the reference library only supports pathname to generate a string tho. https://github.com/pillarjs/path-to-regexp#compile-reverse-path-to-regexp
> 
> I'm wondering how we should handle the case if there are two named groups which have the same name?

Yeah, applying it other than pathnames sounds challenging.
Is it realistic to limit the generate application only within the pathname field? 

> const pattern = new URLPattern('https://{:pattern.}?example.com/:pattern');
> const result = await pattern.generate({'pattern': 'image/a.jpg'});
> // => ???
> It looks the current design doesn't have a capability to substitute two strings which have same names in a different component.

Even if it is useless, I feel it natural to generate `{'protocol': 'https', 'hostname': 'image/a.jpgexample.com', 'pathname': /image/a.jpg`,...}`.  It should be a web developer's responsibility to avoid practically unusable cases.

**Comment by annevk** [comment]
Maybe duplicate names should throw? But I was wondering if we could avoid having to specify the component. At least it seems  like it would be nice if you could do
```
const pattern = new URLPattern("https://{:host}/{:path1}/{:path2}/");
pattern.generate({ host: "example.com", path1: "foo", path2: "bar/baz" });
// "https://example.com/foo/bar%2fbaz/"

@garretrieger maybe you could also weigh in with regards to requirements?

**Comment by garretrieger** [comment]
For the IFT use case the set of variables is fixed and have well defined values (base32/base64 encodings) so it would pretty unusual for IFT template to substitute one of the variables outside of the path or query string.

**Comment by azaika** [comment]
Adding to @sisidovski , we published Blink [Intent to Prototype](https://groups.google.com/a/chromium.org/d/msgid/blink-dev/6837c0a6.170a0220.fb24a.1185.GAE%40google.com) today.
Right now, Chrome only implements the most basic URLPattern syntax: Full-Wildcards, RegExp, and modifiers are out of the initial scope. We would greatly appreciate feedback from web developers to expand the supported features.

**Comment by domenic** [comment]
I agree with both of your intuitions. I guess the second example is hard though if you are not yet doing full-wildcards.

**Comment by azaika** [comment]
Thanks! As of full-wildcards, I have a slight concern that I have also mentioned in [a DesignDoc comment](https://docs.google.com/document/d/1ca6geyHD40MfHkalgEv4AcBo9-rHK6CrAOxwcWor9WA/edit?usp=drivesdk&disco=AAABfmqmQtI): using numeric name to specify full-wildcards may be confusing.
For example, the full-wildcard in `/(abc)/*` has the numeric name `1` per implicit naming in the parsing step ([add-a-part algorithm](https://urlpattern.spec.whatwg.org/#add-a-part)), but some people expect it to have the name `0`.

**Comment by domenic** [comment]
I see, because `(abc)` gets position 0, right?

Let me recap my understanding of the current situation:

1. `generate()` on a URLPattern with segment-wildcard parts and non-delimiter inputs will work
1. `generate()` on a URLPattern with segment-wildcard parts and delimiter inputs will throw
1. `generate()` on a URLPattern with regexp/full-wildcard parts and non-delimiter inputs will throw
1. `generate()` on a URLPattern with regexp/full-wildcard parts and delimiter inputs will throw

If this is correct, I think changing (2) from throw to encode is fine. And we retain optionality on how to change the behavior of (3) and (4) in the future.

**Comment by azaika** [comment]
@domenic You are right. `(abc)` gets position 0, and the current situation is exactly as you summed up. (Sorry if I confused you.)

I am for your idea and still need ideas on (3) and (4). Thanks for sorting out!

**Comment by jyasskin** [comment]
From https://github.com/w3ctag/design-reviews/discussions/1123#discussioncomment-13861273, it looks like MapML might be a use case for this. They're currently using URL Templates (https://www.rfc-editor.org/rfc/rfc6570.html I think), and it'd be good to talk to @prushforth to see if we can ensure that markup attributes can use a single templating format for both matching and generation.

**Comment by prushforth** [comment]
It would be great if there was a platform primitive for substituting values into URLs that could be used by HTML elements and scripting applications alike - off the top of my head, similar to how `media` attribute and `matchMedia` API may work (I think). 

I think it could be useful for applications other than maps.  But it's definitely useful in maps for embedding fetches for tiles, images and other 2D-covering resources.  Maybe it would be useful in 3D as well, who knows.  Happy to discuss anytime.

**Comment by azaika** [comment]
> I wonder if the presence of ":" in "EPSG:3857" in the above hrefpattern would be problematic? It doesn't represent a variable for substitution, it's part of the string content of the URL that's required by the server's CRS parameter value.

Yes, IIUC, the ":" in strings should be escaped. For example, `"EPSG:3857"` should be encoded as `"EPSG\\:3857"` for URLPattern.

**Comment by uncenter** [comment]
I'm a little confused because I'm seeing two different use cases that don't use the same API here. One use case is to *execute* a `URLPattern` against a URL, modify the `URLPatternResult`, and (re)serialize/stringify the result back into a URL.

This is what @jimmywarting discussed.

> I were just merely considering if a result could be mutable... for example: Say that you are on `shopify.com/categories/computers` and want to navigate to games section `shopify.com/categories/games`
> 
> const pattern = new URLPattern({ pathname: ':vendor/categories/:category' })
> // Get current url pattern
> const result = pattern.exec(location.href)
> 
> result.pathname.groups.category = 'games' // Modify current url match result
> location.href = result.toString() // change the url location
> 
> Thinking this could be useful for a adv search with 2 way data binding or something.
> 
> all doe a `generate` function could work as equally good, maybe mutable url pattern result isn't desirable?

The second use case is to take a URLPattern and fill in the pattern's parameters to make a URL or URL component. This is what @azaika and @sisidovski cover in the proposed https://github.com/explainers-by-googlers/urlpattern-generate API. *This is also I believe the original focus of this thread, as per the starting comment:*

> One of the most frequent pieces of feedback we have gotten is that developers would like to generate URL and component strings from a URLPattern. For example, something like:
> 
> const pattern = new URLPattern({ pathname: '/product/:id' });
> const my_pathname = pattern.generate('pathname', { id: '12345' });
> // my_pathname is "/product/12345"

Would it make sense to create a new issue here to split out the first use case about modifying and serializing `URLPatternResult`s? Please correct me if I'm missing some part of this thread :)

**Comment by sisidovski** [comment]
I think you are correct that from a developer's perspective, both approaches serve the same high-level goal: obtaining a URL string from a pattern.

Rather than seeing them as two different use cases, I view them as two distinct approaches to solve that single goal.

* Stateless Generation: Filling in parameters directly via URLPattern.generate().
* Stateful Transformation: Executing a pattern, modifying the URLPatternResult, and re-serializing it.

As I mentioned earlier, the second approach is technically more challenging because URLPatternResult is currently defined as a dictionary and lacks methods.

But do you have any concrete use case that doesn't work with the stateless approach but work with the stateful approach?

**Comment by uncenter** [comment]
How would the example I quoted for the first use case be solved by the proposed `generate' API?

**Comment by sisidovski** [comment]
You can solve this by passing the existing match results into the generate() method. Simply we use the spread operator to override the specific parameters you want to change:

```js
const pattern = new URLPattern({ pathname: ':vendor/categories/:category' });
const result = pattern.exec(location.href);
const newPathname = pattern.generate('pathname', {
  ...result.pathname.groups,
  category: 'games'
});
location.pathname = newPathname;
```
